File: PES-U/beneficiary-dedup-engine/python/algorithms/levenshtein.py
"""
algorithms/levenshtein.py   <->   lib/src/algorithms/levenshtein.dart

Levenshtein edit distance and a normalized [0,1] similarity, plus the
Damerau variant (adds adjacent transposition as a single edit).

  levenshtein_distance("ibrahim", "ibrahmi") -> 2   (plain Levenshtein)
  damerau_distance   ("ibrahim", "ibrahmi") -> 1   (one transposition)

Damerau is better for name typos because swapped adjacent letters are common.

Dart port note: these are classic DP tables. Use List<List<int>> or two
rolling List<int> rows. No external packages needed.
"""


def levenshtein_distance(a: str, b: str) -> int:
    """Classic Levenshtein edit distance (insert/delete/substitute)."""
    if a == b:
        return 0
    if len(a) == 0:
        return len(b)
    if len(b) == 0:
        return len(a)

    # Two-row rolling DP for O(min(m,n)) memory — easy to port.
    previous = list(range(len(b) + 1))
    current = [0] * (len(b) + 1)

    for i in range(1, len(a) + 1):
        current[0] = i
        for j in range(1, len(b) + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            current[j] = min(
                previous[j] + 1,        # deletion
                current[j - 1] + 1,     # insertion
                previous[j - 1] + cost, # substitution
            )
        previous, current = current, previous

    return previous[len(b)]


def damerau_levenshtein_distance(a: str, b: str) -> int:
    """
    Damerau-Levenshtein (optimal string alignment variant).
    Like Levenshtein but a swap of two adjacent chars costs 1, not 2.

    Dart port note: needs the full (m+1)x(n+1) matrix because of the
    transposition look-back. Use List<List<int>>.
    """
    if a == b:
        return 0
    len_a = len(a)
    len_b = len(b)
    if len_a == 0:
        return len_b
    if len_b == 0:
        return len_a

    d = [[0] * (len_b + 1) for _ in range(len_a + 1)]
    for i in range(len_a + 1):
        d[i][0] = i
    for j in range(len_b + 1):
        d[0][j] = j

    for i in range(1, len_a + 1):
        for j in range(1, len_b + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            d[i][j] = min(
                d[i - 1][j] + 1,         # deletion
                d[i][j - 1] + 1,         # insertion
                d[i - 1][j - 1] + cost,  # substitution
            )
            # Transposition
            if (i > 1 and j > 1
                    and a[i - 1] == b[j - 2]
                    and a[i - 2] == b[j - 1]):
                d[i][j] = min(d[i][j], d[i - 2][j - 2] + 1)

    return d[len_a][len_b]


def levenshtein_similarity(a: str, b: str) -> float:
    """Normalized: 1 - distance / max_len. Returns [0,1]."""
    if not a and not b:
        return 1.0
    if not a or not b:
        return 0.0
    max_len = max(len(a), len(b))
    return 1.0 - levenshtein_distance(a, b) / max_len


def damerau_similarity(a: str, b: str) -> float:
    """Normalized Damerau-Levenshtein similarity. Returns [0,1]."""
    if not a and not b:
        return 1.0
    if not a or not b:
        return 0.0
    max_len = max(len(a), len(b))
    return 1.0 - damerau_levenshtein_distance(a, b) / max_len

File: PES-U/beneficiary-dedup-engine/dedup_engine/lib/src/algorithms/levenshtein.dart
/// Levenshtein and Damerau-Levenshtein edit distance, plus normalized
/// [0,1] similarity scores.
///
/// Damerau counts a swap of two adjacent characters as ONE edit, which is
/// what a typo like "ibrahmi" for "ibrahim" actually is.

/// Classic Levenshtein edit distance (insert / delete / substitute).
library; 
int levenshteinDistance(String a, String b) {
  if (a == b) return 0;
  if (a.isEmpty) return b.length;
  if (b.isEmpty) return a.length;

  // Two rolling rows — O(min(m,n)) memory.
  var previous = List<int>.generate(b.length + 1, (i) => i);
  var current = List<int>.filled(b.length + 1, 0);

  for (var i = 1; i <= a.length; i++) {
    current[0] = i;
    for (var j = 1; j <= b.length; j++) {
      final cost = (a[i - 1] == b[j - 1]) ? 0 : 1;
      final deletion = previous[j] + 1;
      final insertion = current[j - 1] + 1;
      final substitution = previous[j - 1] + cost;
      current[j] = [deletion, insertion, substitution]
          .reduce((x, y) => x < y ? x : y);
    }
    // swap the rows
    final tmp = previous;
    previous = current;
    current = tmp;
  }

  return previous[b.length];
}

/// Damerau-Levenshtein (optimal string alignment): like Levenshtein, but a
/// transposition of two adjacent characters costs 1 instead of 2.
int damerauDistance(String a, String b) {
  if (a == b) return 0;
  if (a.isEmpty) return b.length;
  if (b.isEmpty) return a.length;

  // Full matrix is needed here because of the transposition look-back.
  final d = List<List<int>>.generate(
    a.length + 1,
    (_) => List<int>.filled(b.length + 1, 0),
  );

  for (var i = 0; i <= a.length; i++) {
    d[i][0] = i;
  }
  for (var j = 0; j <= b.length; j++) {
    d[0][j] = j;
  }

  for (var i = 1; i <= a.length; i++) {
    for (var j = 1; j <= b.length; j++) {
      final cost = (a[i - 1] == b[j - 1]) ? 0 : 1;
      var best = [
        d[i - 1][j] + 1, // deletion
        d[i][j - 1] + 1, // insertion
        d[i - 1][j - 1] + cost, // substitution
      ].reduce((x, y) => x < y ? x : y);

      // transposition
      if (i > 1 && j > 1 && a[i - 1] == b[j - 2] && a[i - 2] == b[j - 1]) {
        final transposition = d[i - 2][j - 2] + 1;
        if (transposition < best) best = transposition;
      }

      d[i][j] = best;
    }
  }

  return d[a.length][b.length];
}

/// Normalized Levenshtein similarity in [0.0, 1.0].
double levenshteinSimilarity(String a, String b) {
  if (a.isEmpty && b.isEmpty) return 1.0;
  if (a.isEmpty || b.isEmpty) return 0.0;
  final maxLen = a.length > b.length ? a.length : b.length;
  return 1.0 - levenshteinDistance(a, b) / maxLen;
}

/// Normalized Damerau-Levenshtein similarity in [0.0, 1.0].
double damerauSimilarity(String a, String b) {
  if (a.isEmpty && b.isEmpty) return 1.0;
  if (a.isEmpty || b.isEmpty) return 0.0;
  final maxLen = a.length > b.length ? a.length : b.length;
  return 1.0 - damerauDistance(a, b) / maxLen;
}